libsolidity/parsing/Parser.cpp

```diff
@@ -333,9 +336,17 @@ Parser::FunctionHeaderParserResult Parser::parseFunctionHeader(bool _forceEmptyN
 {
 	RecursionGuard recursionGuard(*this);
 	FunctionHeaderParserResult result;
-	expectToken(Token::Function);
-	if (_forceEmptyName || m_scanner->currentToken() == Token::LParen)
-		result.name = make_shared<ASTString>(); // anonymous function
+
+	if (m_scanner->currentToken() == Token::Function)
+		result.isConstructor = false;
+	else if (m_scanner->currentToken() == Token::Identifier && m_scanner->currentLiteral() == "constructor")
+		result.isConstructor = true;
+	else
+		solAssert(false, "Function or constructor expected.");
```

Comment:
`expectToken` actually results in `fatalParserError`, which is different to `solAssert`. It is strange we didn't had a test for this to catch the discrepancy.

Reply:
Actually expectToken(Token::Function) could never fail here - parseContractDefinition already checked the tokens and only called parseFunctionDefinitionOrFunctionTypeStateVariable (which further calls parseFunctionHeader) if the token was Token::Function anyways.

The same is true now for the token either being Token::Function or "constructor" as Identifier.

For this reason I think "solAssert" is actually more appropriate than fatalParserError.

Reply:
Okay double checked based on your comment and in both cases the precondition is checked. Should be fine.